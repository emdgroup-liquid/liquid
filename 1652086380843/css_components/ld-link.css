.ld-link{--ld-link-col:var(--ld-thm-primary);--ld-link-chevron-width:0.15em;align-items:baseline;color:var(--ld-link-col);cursor:pointer;font-family:var(--ld-font-body);font-size:inherit;font-weight:700;gap:.3em;line-height:inherit;text-decoration:none;touch-action:manipulation}.ld-link:not([aria-disabled=true]):hover{--ld-link-col:var(--ld-thm-primary-hover)}.ld-link:not([aria-disabled=true]):focus:focus-visible{--ld-link-col:var(--ld-thm-primary-focus)}.ld-link:not([aria-disabled=true]):active{--ld-link-col:var(--ld-thm-primary-active)}.ld-link:where([aria-disabled=true]){--ld-link-col:var(--ld-thm-primary);cursor:default;opacity:.3}.ld-link--chevron-end,.ld-link--chevron-start{display:inline-flex}.ld-link--chevron-end:after,.ld-link--chevron-end:before,.ld-link--chevron-start:after,.ld-link--chevron-start:before{background-color:var(--ld-link-col);border-radius:var(--ld-link-chevron-width);content:"";display:block;flex-shrink:0;height:.5em;transform-origin:right;width:var(--ld-link-chevron-width)}.ld-link--chevron-end:before,.ld-link--chevron-start:before{transform:translateX(calc(.15em + 50%)) translateY(-60%) rotate(-40deg)}.ld-link--chevron-end:after,.ld-link--chevron-start:after{transform:translateX(calc(-.15em - 50%)) translateY(20%) rotate(40deg)}.ld-link--chevron-start:after,.ld-link--chevron-start:before{order:-1}.ld-link--chevron-end:after,.ld-link--chevron-end:before{order:1}