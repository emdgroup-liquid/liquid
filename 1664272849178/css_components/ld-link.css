.ld-link{--ld-link-col:var(--ld-thm-primary);--ld-link-gap:0.3em;--ld-link-chevron-gap:0em;--ld-link-chevron-width:0.135em;align-items:baseline;color:var(--ld-link-col);cursor:pointer;font-family:var(--ld-font-body);font-size:inherit;font-weight:700;gap:var(--ld-link-gap);line-height:inherit;text-decoration:none;touch-action:manipulation}.ld-link:not([aria-disabled=true]):hover{--ld-link-col:var(--ld-thm-primary-hover)}.ld-link:not([aria-disabled=true]):focus:focus-visible{--ld-link-col:var(--ld-thm-primary-focus)}.ld-link:not([aria-disabled=true]):active{--ld-link-col:var(--ld-thm-primary-active)}.ld-link:where([aria-disabled=true]){--ld-link-col:var(--ld-thm-primary);cursor:default;opacity:.3}.ld-link>.ld-icon,.ld-link>::slotted(.ld-icon),.ld-link>::slotted(ld-icon),.ld-link>ld-icon{transform:translateY(10%)}.ld-link--chevron-end,.ld-link--chevron-start{display:inline-flex}.ld-link--chevron-end:after,.ld-link--chevron-end:before,.ld-link--chevron-start:after,.ld-link--chevron-start:before{background-color:var(--ld-link-col);border-radius:99rem;content:"";display:block;flex-shrink:0;height:.5em;transform-origin:right;width:var(--ld-link-chevron-width)}.ld-link--chevron-start{margin-left:calc(var(--ld-link-chevron-gap)*.5)}.ld-link--chevron-start:after,.ld-link--chevron-start:before{order:-1}.ld-link--chevron-start:before{transform:translateX(calc((var(--ld-link-gap) - var(--ld-link-chevron-gap))*.5 + 50%)) translateY(-65%) rotate(-40deg)}.ld-link--chevron-start:after{transform:translateX(calc(var(--ld-link-gap)*-.5 - var(--ld-link-chevron-gap)*.5 - 50%)) translateY(20%) rotate(40deg)}.ld-link--chevron-end{margin-right:calc(var(--ld-link-chevron-gap)*.5)}.ld-link--chevron-end:after,.ld-link--chevron-end:before{order:1}.ld-link--chevron-end:before{transform:translateX(calc((var(--ld-link-gap) + var(--ld-link-chevron-gap))*.5 + 50%)) translateY(-65%) rotate(-40deg)}.ld-link--chevron-end:after{transform:translateX(calc(var(--ld-link-gap)*-.5 + var(--ld-link-chevron-gap)*.5 - 50%)) translateY(20%) rotate(40deg)}