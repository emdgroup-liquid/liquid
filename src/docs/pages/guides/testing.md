---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

