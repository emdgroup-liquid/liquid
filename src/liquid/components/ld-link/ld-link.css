.ld-link {
  --ld-link-font-size: 1rem;
  --ld-link-size-sm: 0.875rem;
  --ld-link-size-lg: 1.125rem;
  --ld-typo-link-brand-color: var(--ld-thm-primary);

  position: relative;
  display: inline-block;
  color: var(--ld-typo-link-brand-color);
  font-size: var(--ld-link-font-size);
  font-weight: 700;
  text-decoration: none;
  
  &--sm {
    --ld-link-font-size: var(--ld-link-size-sm);
  }
  
  &--lg {
    --ld-link-font-size: var(--ld-link-size-lg);
  }

  &:hover {
    color: var(--ld-thm-primary-hover);
    cursor: pointer;
  }

  &:focus {
    color: var(--ld-thm-primary-focus);
  }

  &:focus-visible {
    color: var(--ld-thm-primary-focus);
    outline: none;
  }

  &:visited,
  &:active {
    color: var(--ld-thm-primary-active);
  }

  &--disabled {
    opacity: 0.3;
    pointer-events: none;
    color: var(--ld-typo-link-brand-color);
  }
}

.ld-link__icon {
  --ld-icon-size-sm: 0.625rem;
  --ld-icon-size-md: 0.750rem;
  --ld-icon-size-lg: 0.875rem;
}