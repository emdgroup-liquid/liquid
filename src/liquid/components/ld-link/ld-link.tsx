import '../../components' // type definitions for type checks and intelliSense
import { Component, Host, h, Prop } from '@stencil/core'
import { getClassNames } from 'src/liquid/utils/getClassNames'

/**
 * @virtualProp ref - reference to component
 * @virtualProp {string | number} key - for tracking the node's identity when working with lists
 * @part arrow - adds chevron arrow before link text
 * @part size - sets the size of the text
 * @part href - sets the link address
 * @part disabled - sets the disabled state
 * @part target - sets target of the link
 * @part showIcon - displays chevron icon as prefix
 */

@Component({
  tag: 'ld-link',
  styleUrl: 'ld-link.css',
  shadow: true,
})
export class LdLink {
  /** Sets the size of the text */
  @Prop() size?: 'sm' | 'lg'

  /** Sets the link address */
  @Prop() href?: string

  /** Sets the disabled state */
  @Prop() disabled?: boolean

  /** Sets target of the link - _blank|_self|_parent|_top|framename */
  @Prop() target?: string

  /** Displays chevron icon as prefix */
  @Prop() showIcon?: boolean

  render() {
    return (
      <Host>
        <a
          href={this.href}
          target={this.target ?? '_self'}
          class={getClassNames([
            'ld-link',
            this.size && `ld-link--${this.size}`,
            this.disabled && `ld-link--disabled`,
          ])}
        >
          {this.showIcon && (
            <ld-icon
              class="ld-link__icon"
              name="arrow-right"
              part="icon"
              size={this.size}
            />
          )}
          <slot></slot>
        </a>
      </Host>
    )
  }
}
