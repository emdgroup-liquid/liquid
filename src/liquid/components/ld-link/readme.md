---
eleventyNavigation:
  key: Link
  parent: Components
layout: layout.njk
title: Link
permalink: components/ld-link/
---

# ld-link

<link rel="stylesheet" href="css_components/ld-link.css">
<link rel="stylesheet" href="css_components/ld-icon.css">

This component is meant to be used in conjunction with the [`ld-icon`](components/ld-icon/). Please reffer to the [`ld-icon`](components/ld-icon/) docs for further usage examples.

## Examples

### Basic

{% example %}
<ld-link>Link</ld-link>

<!-- CSS component -->

<a class="ld-link">Link</a>

{% endexample %}

### Different sizes

{% example %}
<ld-link size="sm">Link</ld-link>
<ld-link>Link</ld-link>
<ld-link size="lg">Link</ld-link>

<!-- CSS component -->

<a class="ld-link ld-link--sm">Link</a>
<a class="ld-link">Link</a>
<a class="ld-link ld-link--lg">Link</a>

{% endexample %}

### Disabled state

{% example %}
<ld-link disabled>Link</ld-link>

<!-- CSS component -->

<a class="ld-link ld-link--disabled">Link</a>

{% endexample %}

### With link to new address

{% example %}
<ld-link href="introduction/getting-started/">Link</ld-link>

<!-- CSS component -->

<a href="introduction/getting-started/" class="ld-link">Link</a>

{% endexample %}

### With _blank target

{% example %}
<ld-link target="_blank" href="introduction/getting-started/">Link</ld-link>

<!-- CSS component -->

<a target="_blank" href="introduction/getting-started/" class="ld-link">Link</a>

{% endexample %}

### With chevron icon

{% example %}
<ld-link show-icon>Link</ld-link>

<!-- CSS component -->

<a class="ld-link">
<svg class="ld-link__icon ld-icon" viewBox="0 0 16 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="m7 13 4-5-4-5" stroke="currentcolor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round"/>
</svg>Link</a>

{% endexample %}

<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description                                                       | Type               | Default     |
| ---------- | ----------- | ----------------------------------------------------------------- | ------------------ | ----------- |
| `disabled` | `disabled`  | Sets the disabled state                                           | `boolean`          | `undefined` |
| `href`     | `href`      | Sets the link address                                             | `string`           | `undefined` |
| `key`      | `key`       | for tracking the node's identity when working with lists          | `string \| number` | `undefined` |
| `ref`      | `ref`       | reference to component                                            | `any`              | `undefined` |
| `showIcon` | `show-icon` | Displays chevron icon as prefix                                   | `boolean`          | `undefined` |
| `size`     | `size`      | Sets the size of the text                                         | `"lg" \| "sm"`     | `undefined` |
| `target`   | `target`    | Sets target of the link - _blank\|_self\|_parent\|_top\|framename | `string`           | `undefined` |


## Shadow Parts

| Part         | Description                         |
| ------------ | ----------------------------------- |
| `"arrow"`    | adds chevron arrow before link text |
| `"disabled"` | sets the disabled state             |
| `"href"`     | sets the link address               |
| `"icon"`     |                                     |
| `"showIcon"` | displays chevron icon as prefix     |
| `"size"`     | sets the size of the text           |
| `"target"`   | sets target of the link             |


## Dependencies

### Depends on

- [ld-icon](../ld-icon)

### Graph
```mermaid
graph TD;
  ld-link --> ld-icon
  style ld-link fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

 
