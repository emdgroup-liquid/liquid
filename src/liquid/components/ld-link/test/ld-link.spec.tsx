import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdLink } from '../ld-link'

describe('ld-notice', () => {
  it('renders default', async () => {
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link>Link</ld-link>,
    })
    expect(page.root).toMatchSnapshot()
  })

  it('renders with small size', async () => {
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link size="sm">Link</ld-link>,
    })

    const smallElement = page.root.shadowRoot.querySelector('.ld-link--sm')
    expect(smallElement).not.toBeNull()
    expect(page.root).toMatchSnapshot()
  })

  it('renders with large size', async () => {
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link size="lg">Link</ld-link>,
    })

    const largeElement = page.root.shadowRoot.querySelector('.ld-link--lg')
    expect(largeElement).not.toBeNull()
    expect(page.root).toMatchSnapshot()
  })

  it('renders as disabled', async () => {
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link disabled>Link</ld-link>,
    })

    const disabledElement =
      page.root.shadowRoot.querySelector('.ld-link--disabled')
    expect(disabledElement).not.toBeNull()
    expect(page.root).toMatchSnapshot()
  })

  it('renders with chevron icon', async () => {
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link show-icon>Link</ld-link>,
    })

    const icon = page.root.shadowRoot.querySelector<HTMLElement>('ld-icon')
    expect(icon).not.toBeNull()
    expect(page.root).toMatchSnapshot()
  })

  it('renders with href', async () => {
    const url = 'http://testing.stenciljs.com/introduction/getting-started/'
    const page = await newSpecPage({
      components: [LdLink],
      template: () => <ld-link href={url}>Link</ld-link>,
    })

    const anchorElement = page.root.shadowRoot.querySelector<HTMLElement>('a')
    expect(anchorElement).toHaveProperty('href', url)
    expect(page.root).toMatchSnapshot()
  })

  it('renders with target _blank', async () => {
    const url = 'http://testing.stenciljs.com/introduction/getting-started/'
    const target = '_blank'
    const page = await newSpecPage({
      components: [LdLink],
      template: () => (
        <ld-link href={url} target={target}>
          Link
        </ld-link>
      ),
    })

    const anchorElement = page.root.shadowRoot.querySelector<HTMLElement>('a')
    expect(anchorElement).toEqualAttribute('target', target)
    expect(page.root).toMatchSnapshot()
  })
})
