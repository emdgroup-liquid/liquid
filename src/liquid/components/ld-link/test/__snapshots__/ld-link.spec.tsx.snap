// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-notice renders as disabled 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link ld-link--disabled" target="_self">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;

exports[`ld-notice renders default 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link" target="_self">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Link
</ld-link>
`;

exports[`ld-notice renders with chevron icon 1`] = `
<ld-link show-icon="">
  <mock:shadow-root>
    <a class="ld-link" target="_self">
      <ld-icon class="ld-link__icon" name="arrow-right" part="icon"></ld-icon>
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;

exports[`ld-notice renders with href 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link" href="http://testing.stenciljs.com/introduction/getting-started/" target="_self">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;

exports[`ld-notice renders with large size 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link ld-link--lg" target="_self">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;

exports[`ld-notice renders with small size 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link ld-link--sm" target="_self">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;

exports[`ld-notice renders with target _blank 1`] = `
<ld-link>
  <mock:shadow-root>
    <a class="ld-link" href="http://testing.stenciljs.com/introduction/getting-started/" target="_blank">
      <slot></slot>
    </a>
  </mock:shadow-root>
  Lorem ipsum dolor sit amet.
</ld-link>
`;
